test/e2e/validation.go

```diff
@@ -22,199 +22,211 @@ import (
-			invalidKFC.Spec.APIEndpoint = "https://"
-			err := client.Create(context.TODO(), invalidKFC)
-			if err != nil {
-				f.Logger().Fatalf("Unexpected error creating invalid %s = %+v in another test namespace %s, err: %v", resourceName, invalidKFC, namespace, err)
+			// Immediately delete the created test resource to avoid errors in
+			// other e2e tests that rely on the original e2e testing setup. For
+			// example for KubeFedCluster, delete the test cluster we just
+			// created as it's not a properly joined member cluster that's part
+			// of the original e2e test setup.
+			validObjName := vrt.getObjectMeta().Name
+			err = client.Delete(context.TODO(), validObj, namespace, validObjName)
+			if err != nil && !apierrors.IsNotFound(err) {
+				f.Logger().Errorf("Error deleting %s %s: %v", resourceName, validObjName, err)
 			}
 		})
-	})
+
+		// TODO(font): Consider removing once webhook singleton is implemented.
+		When("running with namespace scoped deployment", func() {
+			It(fmt.Sprintf("for %s should succeed when an invalid %s is created outside the kubefed system namespace", resourceName, resourceName), func() {
+				if !framework.TestContext.LimitedScope {
+					framework.Skipf("Cannot run validation admission webhook namespaced test in a cluster scoped deployment")
+				}
+				userAgent := fmt.Sprintf("test-%s-validation", resourceName)
+				client := f.Client(userAgent)
+				kubeClient := f.KubeClient(fmt.Sprintf("%s-create-namespace", testBaseName))
+				namespace := framework.CreateTestNamespace(kubeClient, testBaseName)
+				framework.AddCleanupAction(func() {
+					framework.DeleteNamespace(kubeClient, namespace)
+				})
+
+				By(fmt.Sprintf("Creating an invalid %s in the separate test namespace %s", resourceName, namespace))
+				invalidObj := vrt.invalidObject(namespace)
+				err := client.Create(context.TODO(), invalidObj)
+				if err != nil {
+					f.Logger().Fatalf("Unexpected error creating invalid %s = %+v in another test namespace %s, err: %v", resourceName, invalidObj, namespace, err)
+				}
+			})
+		})
+	}
 })
+
+type validationResourceTest interface {
+	initialize() pkgruntime.Object
+	getObjectMeta() *metav1.ObjectMeta
+	invalidObject(namespace string) pkgruntime.Object
+	setInvalidField(obj pkgruntime.Object)
+	validObject() pkgruntime.Object
+	invalidObjectFromValid(obj pkgruntime.Object) pkgruntime.Object
+}
+
+type ftcValidationTest struct {
+	object pkgruntime.Object
+}
+
+type kfclusterValidationTest struct {
+	object pkgruntime.Object
+}
+
+func newValidationResourceTest(resourceName string) validationResourceTest {
+	var vrt validationResourceTest
+	switch resourceName {
+	case ftc.ResourceName:
+		vrt = &ftcValidationTest{}
+	case kfcluster.ResourceName:
+		vrt = &kfclusterValidationTest{}
+	}
+	return vrt
+}
+
+func (ftc *ftcValidationTest) initialize() pkgruntime.Object {
+	if ftc.object != nil {
+		return ftc.object
+	}
+
+	// For the target API type, use an existing K8s API resource that
```

Comment:
Could this be simplified to using a type enabled by default if tests were always run in a test namespace or with dry-run?

Reply:
Yes it could, but the namespace-scoped deployment is the only issue. I'm not sure about the dry-run solution for updates per my comment above.